src/libraries/Native/Unix/System.Security.Cryptography.Native/apibridge.c

```diff
@@ -818,4 +818,63 @@ int local_RSA_test_flags(const RSA *r, int flags)
     return r->flags & flags;
 }
 
+int local_EVP_PKEY_check(EVP_PKEY_CTX* ctx)
+{
+    EVP_PKEY* pkey = EVP_PKEY_CTX_get0_pkey(ctx);
+
+    if (pkey == NULL)
+    {
+        ERR_put_error(ERR_LIB_EVP, 0, EVP_R_INPUT_NOT_INITIALIZED, __FILE__, __LINE__);
+        return -1;
+    }
+
+    int id = EVP_PKEY_base_id(pkey);
+
+    switch (id)
+    {
+        case NID_rsaEncryption:
+        {
+            RSA* rsa = EVP_PKEY_get0_RSA(pkey);
+
+            if (rsa != NULL)
+            {
+                return RSA_check_key(rsa);
```

Comment:
This requires "OpenSSL" RSA keys, not keys from an engine. Would this be a problem if the PKEY is from an engine with non-exportable private parameters?

Reply:
It (at least currently) only gets called when importing a PKCS#8 blob into a fresh EVP_PKEY.  So, it'd require

* Running on OpenSSL < 1.1.1 (or we'd call the built-in thing)
* Having a custom ENGINE
* Figuring out how to make that ENGINE be spun up on the d2i_PKCS8_PRIV_KEY_INFO => EVP_PKCS82PKEY pipeline

Seems acceptably marginal.

Once you hit OpenSSL 1.1.1 it's part of the pmeth/ameth function tables, so ENGINEs would be better wired in.  And in OpenSSL 3 the new providers would be even more likely to be properly wired in.

Reply:
> Seems acceptably marginal.

That makes sense.